### PR TITLE
Feat: Abort Requests that takes a lot of time to resolve

### PR DESCRIPTION
**Pull request by Deivu**
**Please describe the changes this PR makes and why it should be merged:**
Lately I suffered from some requests that gets too long to resolve, and due to this, our bot slowed down a lot, as well causing us some problems specially in some of our modules. By adding a request timeout feature for Discord.js requests, we can make sure that requests that takes too long to resolve could get cancelled, hence solving the problem of requests that takes an inconsiderable amount of time to resolve.

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/rest/APIRequest.js [hunk 1]
```diff
@@ -46,11 +47,20 @@ class APIRequest {
...
 
+    const controller = new AbortController();
+    const timeout = setTimeout(() => controller.abort(), 10000);
     return fetch(url, {
       method: this.method,
       headers,
       agent,
       body,
+      signal: controller.signal,
+    }).then(res => {
```
Comment (SpaceEEC) [inline src/rest/APIRequest.js]: I suppose this could use [`Promise#finally`][Promise.prototype.finally], since it seems to be available in node 10, see [node.green].

[Promise.prototype.finally]:https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally
[node.green]:https://node.green/#ES2018-features-Promise-prototype-finally

File: src/rest/APIRequest.js [hunk 2]
```diff
@@ -46,11 +47,20 @@ class APIRequest {
       headers['Content-Type'] = 'application/json';
     }
 
+    const controller = new AbortController();
+    const timeout = setTimeout(() => controller.abort(), 10000);
```
Comment (SpaceEEC) [inline src/rest/APIRequest.js]: The timeout should probably be configurable (or even optional).
Maybe using `ClientOptions`? (Similar to [`retryLimit`])

[`retryLimit`]:https://github.com/discordjs/discord.js/blob/ddcc6cfec9ccb80d42e1d074bf03db2e0dae274e/src/util/Constants.js#L33-L53

**Comment by Deivu** [comment]
Updated as per reviewed.

File: src/rest/APIRequest.js [hunk 3]
```diff
@@ -41,17 +42,20 @@ class APIRequest {
       for (const file of this.options.files) if (file && file.file) body.append(file.name, file.file, file.name);
       if (typeof this.options.data !== 'undefined') body.append('payload_json', JSON.stringify(this.options.data));
       if (!browser) headers = Object.assign(headers, body.getHeaders());
-    } else if (this.options.data != null) { // eslint-disable-line eqeqeq
+    } else if (this.options.data != null) { // eslint-disable-line
```
Comment (Gawdl3y) [inline src/rest/APIRequest.js]: Why was this changed?

Comment (Deivu) [reply src/rest/APIRequest.js]: I thought it was a typo or something, I can just readd it back if I'm wrong

Comment (Monbrey) [reply src/rest/APIRequest.js]: It tells eslint to ignore that the next line is not using strict equality/triple-equals (`eqeqeq`)

Not a typo

Comment (Deivu) [reply src/rest/APIRequest.js]: Thanks for clarification, will update this

File: src/rest/APIRequest.js [hunk 4]
```diff
@@ -46,12 +47,15 @@ class APIRequest {
       headers['Content-Type'] = 'application/json';
     }
 
+    const controller = new AbortController();
+    const timeout = setTimeout(() => controller.abort(), this.client.options.restRequestTimeout);
```
Comment (kyranet) [inline src/rest/APIRequest.js]:
```suggestion
    const timeout = this.client.setTimeout(() => controller.abort(), this.client.options.restRequestTimeout);
```

`this.client` is available at:

https://github.com/discordjs/discord.js/blob/5c0d26bd12ad338280de7ea9626ba52625e0df5c/src/rest/APIRequest.js#L14

And this change allows `Client#destroy()` to clear all the timeouts.

Comment (Deivu) [reply src/rest/APIRequest.js]: Thanks for noticing that, I actually didn't thought about making it as an timeout in client instead. Will be patching things later.

Comment (Deivu) [reply src/rest/APIRequest.js]: PR Updated to use d.js timeout methods instead 👍
